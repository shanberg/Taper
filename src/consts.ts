export const TEMPLATES: Record<string, Template[]> = {
	Default: [
		{ dose: 20, daysForDose: 5 },
		{ dose: 15, daysForDose: 5 },
		{ dose: 10, daysForDose: 5 },
		{ dose: 5, daysForDose: 5 }
	],
	'Giacta 6-month taper': [
		{ dose: 60, daysForDose: 7 },
		{ dose: 50, daysForDose: 7 },
		{ dose: 40, daysForDose: 7 },
		{ dose: 35, daysForDose: 7 },
		{ dose: 30, daysForDose: 7 },
		{ dose: 25, daysForDose: 7 },
		{ dose: 20, daysForDose: 7 },
		{ dose: 15, daysForDose: 7 },
		{ dose: 12.5, daysForDose: 14 },
		{ dose: 10, daysForDose: 7 },
		{ dose: 9, daysForDose: 7 },
		{ dose: 8, daysForDose: 7 },
		{ dose: 7, daysForDose: 7 },
		{ dose: 6, daysForDose: 14 },
		{ dose: 5, daysForDose: 14 },
		{ dose: 4, daysForDose: 14 },
		{ dose: 3, daysForDose: 14 },
		{ dose: 2, daysForDose: 14 },
		{ dose: 1, daysForDose: 14 }
	],
	'Giacta 12-month taper': [
		{ dose: 60, daysForDose: 7 },
		{ dose: 50, daysForDose: 7 },
		{ dose: 40, daysForDose: 7 },
		{ dose: 35, daysForDose: 7 },
		{ dose: 30, daysForDose: 7 },
		{ dose: 25, daysForDose: 7 },
		{ dose: 20, daysForDose: 7 },
		{ dose: 17.5, daysForDose: 14 },
		{ dose: 15, daysForDose: 14 },
		{ dose: 12.5, daysForDose: 7 },
		{ dose: 10, daysForDose: 28 },
		{ dose: 9, daysForDose: 28 },
		{ dose: 8, daysForDose: 28 },
		{ dose: 7, daysForDose: 28 },
		{ dose: 6, daysForDose: 28 },
		{ dose: 5, daysForDose: 28 },
		{ dose: 4, daysForDose: 28 },
		{ dose: 3, daysForDose: 28 },
		{ dose: 2, daysForDose: 28 },
		{ dose: 1, daysForDose: 28 }
	],
	'PMR taper': [
		{ dose: 15, daysForDose: 14 },
		{ dose: 12.5, daysForDose: 14 },
		{ dose: 10, daysForDose: 14 },
		{ dose: 9, daysForDose: 30 },
		{ dose: 8, daysForDose: 30 },
		{ dose: 7, daysForDose: 30 },
		{ dose: 6, daysForDose: 30 },
		{ dose: 5, daysForDose: 90 },
		{ dose: 4, daysForDose: 30 },
		{ dose: 3, daysForDose: 30 },
		{ dose: 2, daysForDose: 30 },
		{ dose: 1, daysForDose: 30 }
	]
};

export const DEFAULT_TEMPLATE_KEY: string = Object.keys(TEMPLATES)[0];

export const LANGUAGES: Language[] = [
	{ labelEn: 'English', lang: 'en-US', verified: true, dir: 'ltr' },
	{ labelEn: 'Spanish', lang: 'es', verified: true, dir: 'ltr' },
	{ labelEn: 'Mandarin', lang: 'zh', verified: false, dir: 'ltr' },
	{ labelEn: 'Haitian Creole', lang: 'ht', verified: false, dir: 'ltr' },
	{ labelEn: 'Swahili', lang: 'sw', verified: false, dir: 'ltr' },
	{ labelEn: 'Arabic', lang: 'ar', verified: false, dir: 'rtl' }
];

export const DEFAULT_LANGUAGE_KEY: string = LANGUAGES[0].lang;

export const PLACEHOLDER_SEGMENT: Segment = { dose: 0, daysForDose: 0 };
