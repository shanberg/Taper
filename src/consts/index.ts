export const TEMPLATES: Record<string, Template[]> = {
	Default: [
		{ dose: 20, duration: 5 },
		{ dose: 15, duration: 5 },
		{ dose: 10, duration: 5 },
		{ dose: 5, duration: 5 }
	],
	'Giacta 6-month taper': [
		{ dose: 60, duration: 7 },
		{ dose: 50, duration: 7 },
		{ dose: 40, duration: 7 },
		{ dose: 35, duration: 7 },
		{ dose: 30, duration: 7 },
		{ dose: 25, duration: 7 },
		{ dose: 20, duration: 7 },
		{ dose: 15, duration: 7 },
		{ dose: 12.5, duration: 14 },
		{ dose: 10, duration: 7 },
		{ dose: 9, duration: 7 },
		{ dose: 8, duration: 7 },
		{ dose: 7, duration: 7 },
		{ dose: 6, duration: 14 },
		{ dose: 5, duration: 14 },
		{ dose: 4, duration: 14 },
		{ dose: 3, duration: 14 },
		{ dose: 2, duration: 14 },
		{ dose: 1, duration: 14 }
	],
	'Giacta 12-month taper': [
		{ dose: 60, duration: 7 },
		{ dose: 50, duration: 7 },
		{ dose: 40, duration: 7 },
		{ dose: 35, duration: 7 },
		{ dose: 30, duration: 7 },
		{ dose: 25, duration: 7 },
		{ dose: 20, duration: 7 },
		{ dose: 17.5, duration: 14 },
		{ dose: 15, duration: 14 },
		{ dose: 12.5, duration: 7 },
		{ dose: 10, duration: 28 },
		{ dose: 9, duration: 28 },
		{ dose: 8, duration: 28 },
		{ dose: 7, duration: 28 },
		{ dose: 6, duration: 28 },
		{ dose: 5, duration: 28 },
		{ dose: 4, duration: 28 },
		{ dose: 3, duration: 28 },
		{ dose: 2, duration: 28 },
		{ dose: 1, duration: 28 }
	],
	'PMR taper': [
		{ dose: 15, duration: 14 },
		{ dose: 12.5, duration: 14 },
		{ dose: 10, duration: 14 },
		{ dose: 9, duration: 30 },
		{ dose: 8, duration: 30 },
		{ dose: 7, duration: 30 },
		{ dose: 6, duration: 30 },
		{ dose: 5, duration: 90 },
		{ dose: 4, duration: 30 },
		{ dose: 3, duration: 30 },
		{ dose: 2, duration: 30 },
		{ dose: 1, duration: 30 }
	]
};

export const DEFAULT_TEMPLATE_KEY: string = Object.keys(TEMPLATES)[0];

export const PLACEHOLDER_STEP: Step = { dose: 0, duration: 0 };
export const PLACEHOLDER_TWICE_STEP: Step = { dose: 0, dose2: 0, duration: 0 };

export const STEP_TYPES = [
	"Daily",
	"Twice Daily",
	"Weekly",
	"Twice Weekly",
] as StepType[]

export const LANGUAGES: Language[] = [
	{ labelEn: 'English', lang: 'en-US', verified: true, dir: 'ltr' },
	{ labelEn: 'Spanish', lang: 'es', verified: true, dir: 'ltr' },
	{ labelEn: 'Mandarin', lang: 'zh', verified: false, dir: 'ltr' },
	{ labelEn: 'Haitian Creole', lang: 'ht', verified: false, dir: 'ltr' },
	{ labelEn: 'Swahili', lang: 'sw', verified: false, dir: 'ltr' },
	{ labelEn: 'Arabic', lang: 'ar', verified: false, dir: 'rtl' }
];

export const DEFAULT_LANGUAGE_KEY: string = LANGUAGES[0].lang;

export const MEDICATIONS: Medication[] = [
	{
		name: "Prednisone",
		unitSizes: [1, 2.5, 5, 10, 20, 50],
		unitType: "solid",
		unit: "mg",
	},
	{
		name: "Gabapentin",
		unitSizes: [100, 300, 400, 600, 800],
		unitType: "solid",
		unit: "mg",
	},
	{
		name: "Clonazepam",
		unitSizes: [0.5, 1, 2],
		unitType: "solid",
		unit: "mg",
	},
	{
		name: "Amitriptyline",
		unitSizes: [10, 25, 50, 75, 100, 150],
		unitType: "solid",
		unit: "mg",
	},
	{
		name: "Sertraline",
		unitSizes: [25, 50, 100],
		unitType: "solid",
		unit: "mg",
	},
	{
		name: "Venlafaxine",
		unitSizes: [37.5, 75, 150, 225],
		unitType: "solid",
		unit: "mg",
	},
	{
		name: "Hydrocortisone",
		unitSizes: [5, 10, 20],
		unitType: "solid",
		unit: "mg",
	},
	{
		name: "Diazepam",
		unitSizes: [2, 5, 10],
		unitType: "solid",
		unit: "mg",
	},
	{
		name: "Methadone",
		unitSizes: [5, 10],
		unitType: "liquid",
		unit: "ml",
	},
	{
		name: "Morphine",
		unitSizes: [15, 30, 60, 100, 200],
		unitType: "liquid",
		unit: "mg",
	}
]