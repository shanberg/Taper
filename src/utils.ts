export function formatDate(date, style = 'conciseUS') {
let options;

  switch (style) {
    case 'conciseUS':
      options = { month: 'numeric', day: 'numeric' };
      break;
    case 'conciseEU':
      options = { day: 'numeric', month: 'numeric', year: '2-digit' };
      break;
    case 'longform':
      options = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' };
      break;
    // Add more styles as needed
    default:
      options = {};
  }

  return new Intl.DateTimeFormat('default', options).format(new Date(date));
}

export const isRowValid = (row): boolean => {
  return row.dose <= 0 || row.daysForDose <= 0
}

export const isRowPlaceholder = (row): boolean => {
  return row.dose === 0 && row.daysForDose === 0
}


// Helper function to format dates as 'YYYY-MM-DD'
export function yyyymmdd(date) {
  const d = new Date(date),
        month = '' + (d.getMonth() + 1),
        day = '' + d.getDate(),
        year = d.getFullYear();

  return [year, month.padStart(2, '0'), day.padStart(2, '0')].join('-');
}