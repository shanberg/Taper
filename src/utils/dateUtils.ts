import { TaperDate } from '../TaperDate';

export const isDateThisYear = (date: ScheduleDate): boolean => {
  return date.getFullYear() === new Date().getFullYear();
};

/** Higher-order function to create a cached formatter */
function createCachedFormatter() {
  const dateTimeFormatCache = new Map();

  return function (date: ScheduleDate, lang: string = 'en-US'): LocaleDate {
    const yearFormat = isDateThisYear(date) ? undefined : 'numeric';
    const cacheKey = `${lang}-${yearFormat}`;

    // Check if the formatter is already in the cache
    if (!dateTimeFormatCache.has(cacheKey)) {
      // Create a new formatter and store it in the cache
      const formatter = new Intl.DateTimeFormat(lang, {
        month: 'short',
        day: 'numeric',
        year: yearFormat
      });
      dateTimeFormatCache.set(cacheKey, formatter);
    }

    // Use the cached formatter
    const formatter = dateTimeFormatCache.get(cacheKey);
    return formatter.format(date) as LocaleDate;
  };
}

// Create a cached version of formatDate
export const cachedFormatDate = createCachedFormatter();


export function calculateSegmentStartAndEndDates(schedule: Schedule, index: number): SegmentWithStartEndDate {
  const segment = schedule.segments[index];
  const taperStartDate = new TaperDate(schedule.startDate);
  const totalDaysForStartDate =
    schedule.segments
      .slice(0, index)
      .reduce((acc: number, curr: { daysForDose: number }) => acc + curr.daysForDose - 1, 0) +
    index;
  taperStartDate.incrementByDays(totalDaysForStartDate);
  const taperEndDate = new TaperDate(taperStartDate.toScheduleDate());
  taperEndDate.incrementByDays(segment.daysForDose - 1);

  return {
    segment,
    segmentStartDate: taperStartDate.toScheduleDate(),
    segmentEndDate: taperEndDate.toScheduleDate()
  }
}
