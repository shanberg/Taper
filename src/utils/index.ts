export * from './dateUtils';
export * from './stepUtils';
export * from './scheduleUtils';
export * from './serializationUtils';
export * from './textUtils';
export * from './languageUtils';