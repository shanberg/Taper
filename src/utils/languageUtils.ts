import { LANGUAGES } from '../consts';

export function getLanguageFromKey(languageKey: string): Language {
  return LANGUAGES.find((l) => l.lang === languageKey) || LANGUAGES[0];
}