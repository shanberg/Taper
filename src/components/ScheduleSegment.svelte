<script lang="ts">
	import { formatSegmentText, isSegmentInvalid } from '../utils';
	import { TaperDate } from '../TaperDate';
	export let segments;
	export let startDate;
	export let selectedLanguage;
	export let segment;
	export let index;

	let segmentStartDate: ScheduleDate;
	let segmentEndDate: ScheduleDate;

	$: isSegmentPlaceholder = segment.dose === 0 && segment.daysForDose === 0;
	$: isInvalid = !isSegmentPlaceholder && isSegmentInvalid(segment);
	$: isLastPlaceholderSegment = index === segments.length - 1 && isSegmentPlaceholder;

	// Calculate the start and end dates
	$: {
		const taperStartDate = new TaperDate(startDate);
		const totalDaysForStartDate =
			segments
				.slice(0, index)
				.reduce((acc: number, curr: { daysForDose: number }) => acc + curr.daysForDose - 1, 0) +
			index;
		taperStartDate.incrementByDays(totalDaysForStartDate);
		const taperEndDate = new TaperDate(taperStartDate.toScheduleDate());
		taperEndDate.incrementByDays(segment.daysForDose - 1);

		segmentStartDate = taperStartDate.toScheduleDate();
		segmentEndDate = taperEndDate.toScheduleDate();
	}
</script>

{#if !isLastPlaceholderSegment}
	<li class:isInvalid class:isSegmentPlaceholder>
		{formatSegmentText({ segment, segmentStartDate, segmentEndDate, index, selectedLanguage })}
	</li>
{/if}

<style>
	.isInvalid {
		color: var(--color-fg-error);
	}
	.isSegmentPlaceholder {
		color: var(--color-fg-muted);
	}
	.isInvalid,
	.isSegmentPlaceholder {
		text-decoration: underline;
		text-decoration-style: wavy;
		text-decoration-skip-ink: none;
	}
</style>
