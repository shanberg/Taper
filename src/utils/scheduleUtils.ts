import { TaperDate } from '../TaperDate';
import { TEMPLATES, DEFAULT_LANGUAGE_KEY, DEFAULT_TEMPLATE_KEY } from '../consts';
import { isSegmentInvalid, sumSegmentsDays, sumSegmentsDose, isSegmentPlaceholder } from './segmentUtils';
import { formatSegmentText } from './textUtils';
import { getLanguageFromKey } from './languageUtils';

const PLACEHOLDER_SEGMENT: Segment = { dose: 0, daysForDose: 0 };

export function createInitialSchedule(): Schedule {
  return {
    segments: [...TEMPLATES.Default, PLACEHOLDER_SEGMENT],
    startDate: new TaperDate().toScheduleDate(),
    templateKey: DEFAULT_TEMPLATE_KEY,
    languageKey: DEFAULT_LANGUAGE_KEY,
    displayMode: "segments"
  };
}

export function isValidSchedule(schedule: Schedule): boolean {
  if (!schedule) {
    throw new Error("No schedule provided");
  }
  return !schedule.segments.slice(0, schedule.segments.length - 1).some(s => isSegmentInvalid(s));
}

export function calculateScheduleSummary(schedule: Schedule): string {
  return `${sumSegmentsDose(schedule.segments)}mg over ${sumSegmentsDays(schedule.segments)} days`
};

export function calculateSegmentStartAndEndDates(schedule: Schedule, index: number): SegmentWithStartEndDate {
  const segment = schedule.segments[index];
  const taperStartDate = new TaperDate(schedule.startDate);
  const totalDaysForStartDate =
    schedule.segments
      .slice(0, index)
      .reduce((acc: number, curr: { daysForDose: number }) => acc + curr.daysForDose - 1, 0) +
    index;
  taperStartDate.incrementByDays(totalDaysForStartDate);
  const taperEndDate = new TaperDate(taperStartDate.toScheduleDate());
  taperEndDate.incrementByDays(segment.daysForDose - 1);

  return {
    segment,
    segmentStartDate: taperStartDate.toScheduleDate(),
    segmentEndDate: taperEndDate.toScheduleDate()
  }
}

export function getFormattedListForCopyPaste(schedule: Schedule): string {
  const { segments, languageKey } = schedule;
  const selectedLanguage = getLanguageFromKey(languageKey);

  // Filter out placeholder segments
  const validSegments = segments.filter(segment => !isSegmentPlaceholder(segment));

  // Format each segment
  const formattedSegments = validSegments.map((segment, index) => {
    const { segmentStartDate, segmentEndDate } = calculateSegmentStartAndEndDates(schedule, index);

    return formatSegmentText({
      segment,
      segmentStartDate,
      segmentEndDate,
      index,
      selectedLanguage
    });
  });

  // Combine formatted segments into a single string
  return formattedSegments.join('\n');
}
