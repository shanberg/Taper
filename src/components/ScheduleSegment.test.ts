import { describe, test, expect, vi, beforeEach, type Mock } from 'vitest';
import { render } from '@testing-library/svelte';
import { LANGUAGES } from '../consts';
import '@testing-library/jest-dom';
import ScheduleSegment from './ScheduleSegment.svelte';

vi.mock('../utils', () => ({
	formatSegmentText: vi.fn(),
	isSegmentInvalid: vi.fn()
}));

import { formatSegmentText, isSegmentInvalid } from '../utils';

describe('ScheduleSegment', () => {
	let segments: Segment[];
	let startDate: Date;
	let selectedLanguage: Language;
	let segment: Segment;
	let index: number;

	beforeEach(() => {
		segments = [
			{ dose: 1, daysForDose: 5 },
			{ dose: 2, daysForDose: 3 }
		];
		startDate = new Date('2024-06-01');
		selectedLanguage = LANGUAGES.find((language) => language.labelEn === 'Arabic') as Language;
		segment = { dose: 1, daysForDose: 5 };
		index = 0;

		(formatSegmentText as Mock).mockImplementation(
			({ segment, segmentStartDate, segmentEndDate, index }) => {
				return `Segment ${index + 1}: ${segment.dose} dose(s) from ${segmentStartDate.toLocaleDateString()} to ${segmentEndDate.toLocaleDateString()}`;
			}
		);

		(isSegmentInvalid as Mock).mockImplementation(
			(segment) => segment.dose < 0 || segment.daysForDose < 0
		);
	});

	test('renders correctly', () => {
		const { container } = render(ScheduleSegment, {
			props: {
				segments,
				startDate,
				selectedLanguage,
				segment,
				index
			}
		});
		expect(container).toMatchSnapshot();
	});

	test('applies isInvalid class when the segment is invalid', () => {
		segment = { dose: -1, daysForDose: 5 };
		const { container } = render(ScheduleSegment, {
			props: {
				segments,
				startDate,
				selectedLanguage,
				segment,
				index
			}
		});
		expect(container.querySelector('li')).toHaveClass('isInvalid');
	});

	test('applies isSegmentPlaceholder class when the segment is a placeholder', () => {
		segment = { dose: 0, daysForDose: 0 };
		const { container } = render(ScheduleSegment, {
			props: {
				segments,
				startDate,
				selectedLanguage,
				segment,
				index
			}
		});
		expect(container.querySelector('li')).toHaveClass('isSegmentPlaceholder');
	});

	test('does not render the last placeholder segment', () => {
		segments = [
			{ dose: 1, daysForDose: 5 },
			{ dose: 0, daysForDose: 0 }
		];
		segment = { dose: 0, daysForDose: 0 };
		index = 1;
		const { container } = render(ScheduleSegment, {
			props: {
				segments,
				startDate,
				selectedLanguage,
				segment,
				index
			}
		});
		expect(container.querySelector('li')).not.toBeInTheDocument();
	});

	test('formats the segment text correctly', () => {
		const { getByText } = render(ScheduleSegment, {
			props: {
				segments,
				startDate,
				selectedLanguage,
				segment,
				index
			}
		});
		expect(getByText(/Segment 1: 1 dose\(s\) from/)).toBeInTheDocument();
	});
});
