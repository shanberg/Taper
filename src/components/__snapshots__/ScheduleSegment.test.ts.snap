// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ScheduleSegment > renders correctly 1`] = `
<div>
  <li
    class="svelte-1eqebjg"
  >
    Segment 1: 1 dose(s) from 6/1/2024 to 6/5/2024
  </li>
  
</div>
`;
